Editing - Nodes orchestration (#3653)
elastic/cloud-on-k8s: docs/orchestrating-elastic-stack-applications/elasticsearch/orchestration.asciidoc

        storageClassName: standard
----

In this example, the Elasticsearch resource defines two NodeSets: 

*  `master-nodes` with 10Gi volumes
*  `data-nodes` with 1000Gi volumes

The Elasticsearch cluster is composed of 13 nodes: 3 master nodes and 10 data nodes.

[id="{p}-upgrading"]
== Upgrading the cluster

ECK handles smooth upgrades from one cluster specification to another. You can apply a new Elasticsearch specification at any time. For example, based on the Elasticsearch specification described in the <<{p}-nodesets,NodeSets overview>>, you can:

> *  Add five additional Elasticsearch data nodes: In `data-nodes` change the value in the `count` field from 10 to 15.

Author: It's hard to say. Might even become heavier. I would leave it as is.

* `podTemplate` can be used to specify  <<{p}-customize-pods,custom settings>> for the Elasticsearch Pod, overriding the default ones set by ECK on the generated StatefulSet specification.
* The StatefulSet name is built from the Elasticsearch resource name and the NodeSet name. Each Pod gets the StatefulSet name suffixed by an ordinal. The corresponding Elasticsearch node has the same name as the Pod.

The actual Pod creation is handled by the StatefulSet controller in Kubernetes. ECK relies on the link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#on-delete[OnDelete StatefulSet update strategy] since it needs full control over when and how Pods get upgraded to a new revision.

When a Pod is removed and recreated (maybe with a newer revision), the StatefulSet controller makes sure that the PersistentVolumes attached to the original Pod are then attached to the new Pod.

[id="{p}-upgrade-patterns"]
== Cluster upgrade patterns

Depending on how the NodeSets are updated, ECK handles the Kubernetes resources reconciliation in various ways.

> * A new NodeSet is added to the Elasticsearch resource.

Author: Same comment above. Might be heavier.

* An existing NodeSet is renamed.
+
ECK creates a new NodeSet with the new name, and removes the old NodeSet, according to the NodeSet creation and removal patterns described above. Elasticsearch data is migrated away from the deprecated NodeSet before removal. The Elasticsearch resource <<{p}-update-strategy,update strategy>> controls how many nodes can exist above or below the target node count during the upgrade.

In all these cases, ECK handles StatefulSet operations according to the Elasticsearch orchestration best practices, by adjusting these orchestration settings accordingly:

*  `discovery.seed_hosts`
*  `cluster.initial_master_nodes`
*  `discovery.zen.minimum_master_nodes`
*  `_cluster/voting_config_exclusions`

[id="{p}-orchestration-limitations"]
== Limitations

> Based on how Kubernetes and StatefulSets operate, ECK orchestration has the following limitations:

Reviewer: ```suggestion
Due to relying on Kubernetes primitives such as StatefulSets, the ECK orchestration process has some inherent limitations.
```

> * Storage requirements (including volume size) of an existing NodeSet cannot be updated. link:https://github.com/kubernetes/enhancements/issues/661[StatefulSet volumes expansion] is not available in Kubernetes yet. To upgrade the storage size, you can create a new NodeSet, or rename an existing one. Renaming a NodeSet automatically creates a new StatefulSet with the specified storage size. The original StatefulSet is removed once the Elasticsearch data is migrated to the nodes of the new StatefulSet.

Reviewer: ```suggestion
* Storage requirements (including volume size) of an existing NodeSet cannot be updated. link:https://github.com/kubernetes/enhancements/issues/661[StatefulSet volumes expansion] is not available in Kubernetes yet. To upgrade the storage size, you can add a new NodeSet or modify an existing one by renaming it first. Changing the name of the NodeSet  triggers the creation of a new StatefulSet which will then inherit the newly updated storage size as well. ECK will ensure that the data from the old StatefulSet is migrated to the new one before the old one is deleted.
```

> * Cluster availability is not guaranteed in the following cases:

Reviewer: ```suggestion
* Cluster availability during rolling upgrades is not guaranteed for the following cases:
```

> ** During the rolling upgrade of single-node clusters
> 
> ** For clusters that have indices with no replicas

Reviewer: ```suggestion
** Single-node clusters
** Clusters containing indices with no replicas
```

If an Elasticsearch node holds the only copy of a shard, this shard becomes unavailable while the node is upgraded. Clusters with more than one node and at least one replica per index are recommended.

* Elasticsearch Pods may stay `Pending` during a rolling upgrade if the Kubernetes scheduler cannot re-schedule them back. This is especially important when using local PersistentVolumes. If the Kubernetes node bound to a local PersistentVolume does not have enough capacity to host an upgraded Pod which was temporarily removed, that Pod will stay `Pending`.

> * Rolling upgrades can make progress if the Elasticsearch cluster health is green. ECK will also make progress if the cluster health is yellow under the following conditions:

Reviewer: ```suggestion
* Rolling upgrades can only make progress if the Elasticsearch cluster health is green. There are exceptions to this rule if the cluster health is yellow and if the following conditions are satisfied:
```

> ** A cluster version upgrade is in progress and some Pods are not up to date.

Reviewer: ```suggestion
** A cluster version upgrade is in progress and some Pods are not up-to-date.
```

Author: Hyphenate it only when used as an adjective => an up-to-date Pod.

** There are no initializing or relocating shards.

If the above conditions are met, then ECK can delete a Pod for upgrade even if the cluster health is yellow, as long as the Pod is not holding the last available replica of a shard.

The health of the cluster is deliberately ignored in the following cases:

> ** Is all the Elasticsearch nodes of a NodeSet are unavailable, probably caused by a misconfiguration, the operator ignores the cluster health and upgrades nodes of the NodeSet.

Reviewer: ```suggestion
** If all the Elasticsearch nodes of a NodeSet are unavailable, probably caused by a misconfiguration, the operator ignores the cluster health and upgrades nodes of the NodeSet.
```

> ** If an Elasticsearch node to upgrade is not healthy, and not part of the Elasticsearch cluster, the operator ignores the cluster health and upgrades the Elasticsearch node.

Reviewer: This line doesn't make sense to me. Is it possible to have an Elasticsearch node that doesn't belong to a cluster?

Author: I guess it was just a typo from my side.
